File: Perl/day02.pl
#!/usr/bin/env perl
BEGIN {
    use Cwd;
    our $directory = cwd;
    our $local_lib = $ENV{"HOME"} . '/perl5/lib/perl5';
}

use lib $directory;
use lib $local_lib;

use Modern::Perl 2018;
use autodie;
use Data::Dumper;
#use Storable 'dclone';

use AOC::Util;
#use AOC::SpatialUtil;

my $INPUT_PATH = '../input';
#my $INPUT_FILE = 'day02_test.txt';
my $INPUT_FILE = 'day02_challenge.txt';
my @input = read_input("$INPUT_PATH/$INPUT_FILE");

say "Advent of Code 2022, Day 02: Rock Paper Scissors";

my @moves = parse_input(@input);

solve_part_one(@moves);
solve_part_two(@moves);

exit( 0 );

sub solve_part_one {
	my @moves = @_;
	
	my %score_lookup = ('X' => 1, 'Y' => 2, 'Z' => 3);
	my $score = 0;
	
	for my $move (@moves) {
		my $move_score = $score_lookup{$move->[1]};
		my $result_score = round_score($move);
		$score += ($move_score + $result_score);
	}
	
	say "Part One: your score is $score.";
}

sub solve_part_two {
	my @input = @_;
}

sub round_score {
	my $move = shift;
	
	# Draw
	if ($move->[0] eq $move->[1]) { return 3; }
	
	if ($move->[1] eq 'X') {
		# Rock X beats Scissors Z
		return $move->[0] eq 'Z' ? 6 : 0;
	}
	elsif ($move->[1] eq 'Y') {
		# Paper Y beats Rock X
		return $move->[0] eq 'X' ? 6 : 0;
	}
	# Scissors Z beats Paper Y
	return $move->[0] eq 'Y' ? 6 : 0;
}

sub parse_input {
	my @lines = @_;
	
	my %translation = ('A' => 'X', 'B' => 'Y', 'C' => 'Z');
	
	my @moves = ();
	for my $line (@lines) {
		my @move = split(' ', $line);
		# Making A,B,C => X,Y,Z
		$move[0] = $translation{$move[0]};
		push(@moves, \@move);
	}
	
	return @moves;
}

File: PHP/day02.php
<?php

require 'AoC/Util.php';
#require 'AoC/SpatialUtil.php';

echo "Advent of Code 2022, Day 02: Rock, Paper, Scissors\n";

$INPUT_DIR = '../input/';
#$INPUT_FILE = 'day02_test.txt';
$INPUT_FILE = 'day02_challenge.txt';
$input = read_input($INPUT_DIR . $INPUT_FILE);

solvePartOne($input);
solvePartTwo($input);


function solvePartOne(array $input) {
	$DRAW = 3; $WIN = 6; $LOSS = 0;
	$score = array( 'A X' => $DRAW + 1,
					'A Y' => $WIN  + 2,
					'A Z' => $LOSS + 3,
					'B X' => $LOSS + 1,
					'B Y' => $DRAW + 2,
					'B Z' => $WIN  + 3,
					'C X' => $WIN + 1,
					'C Y' => $LOSS + 2,
					'C Z' => $DRAW + 3 );
	 $sum = 0;
	 foreach ($input as $line) {
	 	$sum += $score[$line];
	 }
	 
	 print "Part One: score is $sum";
}

function solvePartTwo(array $input) {
	$DRAW = 3; $WIN = 6; $LOSS = 0;
	$score = array( 'A X' => $LOSS + 3,
					'A Y' => $DRAW + 1,
					'A Z' => $WIN  + 2,
					'B X' => $LOSS + 1,
					'B Y' => $DRAW + 2,
					'B Z' => $WIN  + 3,
					'C X' => $LOSS + 2,
					'C Y' => $DRAW + 3,
					'C Z' => $WIN  + 1 );
	 $sum = 0;
	 foreach ($input as $line) {
	 	$sum += $score[$line];
	 }
	 
	 print "Part Two: score is $sum";
}
